### PR TITLE
SNOW-3031774: Add E2E tests for SQL_C_BINARY to numeric, real, decimal, and temporal SQL targets

### PR DESCRIPTION
**Pull request by sfc-gh-makowalski**
## Summary

E2E test coverage for all `SQL_C_BINARY` parameter-binding conversions introduced in PR #833 (numeric / real / decimal / temporal SQL targets), plus a small spec-compliance follow-up to #833 in `odbc/src/conversion/real.rs`.

## Stacked on top of #899

This PR depends on **#899** (`fix: emit "Infinity" / "-Infinity" for non-finite f64 in real bind serialization`).

Without #899, the `should accept SQL_C_BINARY infinity for SQL_REAL` E2E scenario in this PR fails at server SQL-compile time:

```
SQLSTATE=42000: SQL compilation error: Invalid bind value (inf) for type (REAL).
```

Snowflake's JSON parameter-binding parser is stricter than its `TO_DOUBLE()` SQL parser and requires the full word `"Infinity"` / `"-Infinity"` rather than the short `"inf"` Rust's `<f64 as Display>::fmt` produces.

The base branch on this PR is `fix/snowflake-real-json-non-finite-binding`. Once #899 lands on `main`, GitHub will automatically rebase this PR onto `main` and the diff/CI here will reflect the test-only contribution.

## What this PR contributes

- 4 new C++ E2E tests under `odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_*.cpp` covering the full `SQL_C_BINARY -> {DECIMAL, NUMERIC, INTEGER/SHORT/TINYINT/BIGINT, REAL, DOUBLE, DATE, TIME, TIMESTAMP_NTZ}` matrix from #833. 
- 4 matching Gherkin features under `tests/definitions/odbc/types/`.
- 4 new entries in `odbc_tests/tests/e2e/CMakeLists.txt`.
- **BD#46** added to `odbc_tests/BehaviorDifferences.yaml`, scoped per spec analysis to the legitimate gap on the old driver: `SQL_C_BINARY -> SQL_DECIMAL/NUMERIC/TIME/TIMESTAMP` returns SQLSTATE 22003 even for correctly-sized buffers (vs. the new driver's spec-compliant pass-through).
- **`SKIP_OLD_DRIVER("BD#46", …)`** on the 5 tests where the old driver rejects spec-compliant input.
- **Driver fix** in `odbc/src/conversion/real.rs`: drop the `is_finite()` rejection on the `SQL_C_BINARY -> SQL_REAL/DOUBLE` bind path. Per MS ODBC ["C to SQL: Binary"](https://learn.microsoft.com/en-us/sql/odbc/reference/appendixes/c-to-sql-binary) the spec mandates only a length-equals check; the bytes are then passed through. NaN / ±Infinity are valid IEEE-754 values that Snowflake FLOAT columns accept (combined with #899 they now round-trip end-to-end).
- 2 unit tests in `odbc/src/conversion/param_binding.rs` flipped from `_fails` to `_forwards_to_server` to assert NaN serializes as `"NaN"` and Infinity serializes as `"Infinity"` (the strings #899's `write_json` now produces).
- `TEST_CASE_METHOD(ConnSchemaFixture, …)` + `CREATE TEMPORARY TABLE` throughout, matching the established repo idiom on main.

## Validation

- `cargo test --lib`: 973 passed.
- `pre-commit`: rustfmt + clippy + cargo check + clang-format + tests-format-validator all pass.
- Old driver (`odbc_reference_tests` Linux/Windows): green on PR #834 run #7 — confirms the `SKIP_OLD_DRIVER` annotations + test flips are correct.
- New driver: 1 test currently fails (`should accept SQL_C_BINARY infinity for SQL_REAL`) — passes once #899 is part of the base.

## Test plan

- [x] `cargo test --lib`
- [x] `pre-commit run`
- [x] Full `odbc_tests` suite in CI (with #899 in base)
- [x] `odbc_reference_tests` (old driver) stays green

### CODE REVIEW

**Comment by sfc-gh-makowalski** [comment]
* **#834** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/834?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/834?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#833** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/833?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by sfc-gh-makowalski** [comment]
Requesting a fresh Copilot review on the spec-compliance fix. Commit `665a7c2b8` removes the non-spec-compliant NaN/Inf client-side rejection (was a regression introduced in #833) and flips the two corresponding tests to assert round-trip acceptance per MS ODBC *'C to SQL: Binary'* and Snowflake's FLOAT semantics. Tightened BD#46 narrative and SKIP_OLD_DRIVER annotations to only cover the legitimate spec gap (old driver's missing support for SQL_C_BINARY -> DECIMAL/NUMERIC/TIME/TIMESTAMP).

@copilot please review

File: odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_decimal.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#include "get_data.hpp"
+#include "odbc_cast.hpp"
+#include "odbc_matchers.hpp"
+
+static SQL_NUMERIC_STRUCT make_numeric(SQLCHAR precision, SQLSCHAR scale, SQLCHAR sign, uint64_t magnitude) {
+  SQL_NUMERIC_STRUCT ns = {};
+  ns.precision = precision;
+  ns.scale = scale;
+  ns.sign = sign;
+  std::memcpy(ns.val, &magnitude, sizeof(magnitude));
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_decimal.cpp]: Addressed in c4fdd5e9 — populating `SQL_NUMERIC_STRUCT::val` explicitly in little-endian order via byte shifts so the helper is correct on big-endian hosts as well. Thanks for catching this.

File: odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_fixed.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  // Given Snowflake client is logged in
+  Connection conn;
+  auto random_schema = Schema::use_random_schema(conn);
+  conn.execute("CREATE TABLE t (col NUMBER)");
+
+  // When A 4-byte binary buffer containing an i32 is bound as SQL_INTEGER and inserted
+  auto stmt = conn.createStatement();
+  SQLRETURN ret = SQLPrepare(stmt.getHandle(), sqlchar("INSERT INTO t VALUES (?)"), SQL_NTS);
+  REQUIRE_ODBC(ret, stmt);
+  SQLINTEGER val = 42;
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_fixed.cpp:22]: This file already uses `CREATE TEMPORARY TABLE` (line 16) and `TEST_CASE_METHOD(ConnSchemaFixture, ...)`, so each test case runs in its own session-scoped schema and the table is auto-dropped at session end. The "permanent table" concern doesn't apply here.

File: odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_temporal.cpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+// SQL_C_BINARY → SQL_TYPE_DATE
+// ============================================================================
+
+TEST_CASE("should bind SQL_C_BINARY date struct to SQL_TYPE_DATE and read back",
+          "[c_binary][conversion][sql_temporal]") {
+  // Given Snowflake client is logged in
+  Connection conn;
+  auto random_schema = Schema::use_random_schema(conn);
+  conn.execute("CREATE TABLE t (col DATE)");
+
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_temporal.cpp:22]: This file uses `TEST_CASE_METHOD(ConnSchemaFixture, ...)` (line 17 and below) — `Schema::use_random_schema` is not referenced anywhere in the file. Looks like a stale review on an older revision.

File: odbc/src/conversion/real.rs [hunk 4]
```diff
@@ -439,20 +439,17 @@ impl ReadODBC for SnowflakeReal {
                     }
                     .fail();
                 }
-                let v = if expected == 4 {
+                // Per MS ODBC "C to SQL: Binary" spec, the only validation on
+                // SQL_C_BINARY -> SQL_REAL/DOUBLE is the length-equals check
+                // above; the bytes are then interpreted as IEEE-754 and passed
+                // through to the server. NaN and +/-Infinity are valid IEEE-754
+                // values that Snowflake FLOAT columns accept, so we do not
+                // reject them here.
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc/src/conversion/real.rs:482]: The PR description was updated in the same commit (665a7c2b8) that drops the `is_finite()` rejection — it now explicitly says NaN/±Infinity are forwarded to the server per the MS ODBC "C to SQL: Binary" length-equals-only rule, and the matching unit and E2E tests assert acceptance. The metadata is in sync.

File: odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_temporal.cpp [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  // When A 4-byte buffer is bound as SQL_TYPE_TIME
+  auto stmt = conn.createStatement();
+  SQLRETURN ret = SQLPrepare(stmt.getHandle(), sqlchar("INSERT INTO t VALUES (?)"), SQL_NTS);
+  REQUIRE_ODBC(ret, stmt);
+  unsigned char buf[4] = {};
+  SQLLEN ind = sizeof(buf);
+  ret =
+      SQLBindParameter(stmt.getHandle(), 1, SQL_PARAM_INPUT, SQL_C_BINARY, SQL_TYPE_TIME, 0, 0, buf, sizeof(buf), &ind);
+  REQUIRE_ODBC(ret, stmt);
+  ret = SQLExecute(stmt.getHandle());
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_temporal.cpp:101]: The wrong-size SQL_C_BINARY → SQL_TYPE_TIME path is empirically fine on the old driver — `odbc_reference_tests (Linux x64)` and `odbc_reference_tests (Windows x64)` both pass on the current HEAD. BD#46 documents the old driver rejecting at bind for the **correct** struct size (which is why the round-trip tests above have `SKIP_OLD_DRIVER`); the wrong-size path produces 22003 in a way the test still passes, so no skip is needed here.

File: odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_temporal.cpp [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+  // When An 8-byte buffer is bound as SQL_TYPE_TIMESTAMP
+  auto stmt = conn.createStatement();
+  SQLRETURN ret = SQLPrepare(stmt.getHandle(), sqlchar("INSERT INTO t VALUES (?)"), SQL_NTS);
+  REQUIRE_ODBC(ret, stmt);
+  unsigned char buf[8] = {};
+  SQLLEN ind = sizeof(buf);
+  ret = SQLBindParameter(stmt.getHandle(), 1, SQL_PARAM_INPUT, SQL_C_BINARY, SQL_TYPE_TIMESTAMP, 0, 0, buf, sizeof(buf),
+                         &ind);
+  REQUIRE_ODBC(ret, stmt);
+  ret = SQLExecute(stmt.getHandle());
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_temporal.cpp:177]: Same as the SQL_TYPE_TIME wrong-size case — `odbc_reference_tests (Linux/Windows)` are green for the current HEAD, so no `SKIP_OLD_DRIVER` is needed for the wrong-size SQL_TYPE_TIMESTAMP test. The skip annotations are scoped to the round-trip tests above where BD#46 actually bites.

File: odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_decimal.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  // When A 10-byte buffer (not 19) is bound as SQL_DECIMAL
+  auto stmt = conn.createStatement();
+  SQLRETURN ret = SQLPrepare(stmt.getHandle(), sqlchar("INSERT INTO t VALUES (?)"), SQL_NTS);
+  REQUIRE_ODBC(ret, stmt);
+  unsigned char buf[10] = {};
+  SQLLEN ind = sizeof(buf);
+  ret =
+      SQLBindParameter(stmt.getHandle(), 1, SQL_PARAM_INPUT, SQL_C_BINARY, SQL_DECIMAL, 10, 2, buf, sizeof(buf), &ind);
+  REQUIRE_ODBC(ret, stmt);
+  ret = SQLExecute(stmt.getHandle());
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_decimal.cpp:114]: The wrong-size SQL_C_BINARY → SQL_DECIMAL test passes on `odbc_reference_tests (Linux/Windows)` for the current HEAD, so no `SKIP_OLD_DRIVER` is needed here. The skip annotations are correctly scoped to the round-trip DECIMAL/NUMERIC tests above where BD#46 (bind-time 22003 for correctly-sized buffers) actually applies.

File: odbc/src/conversion/param_binding.rs [hunk 8]
```diff
@@ -2891,14 +2897,16 @@ mod tests {
...
+        // Rust's Display for f64::NAN is the literal "NaN" — the server-side
+        // JSON binding parser accepts the same literal for FLOAT targets.
+        assert_eq!(v, Value::String("NaN".to_string()));
+        Ok(())
     }
 
     #[test]
-    fn convert_binary_infinity_to_real_fails() {
+    fn convert_binary_infinity_to_real_forwards_to_server() -> TestResult {
         let val: f32 = f32::INFINITY;
```
Comment (sfc-gh-pbulawa) [inline odbc/src/conversion/param_binding.rs:2966]: Let's test -Infinity too

File: tests/definitions/odbc/types/c_binary_conversion_to_sql_real.feature [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    Then The value should be read back correctly
+
+  @odbc_e2e
+  Scenario: should accept SQL_C_BINARY NaN for SQL_DOUBLE
+    Given Snowflake client is logged in
+    When An 8-byte binary buffer containing NaN is bound as SQL_DOUBLE
+    Then The NaN value should round-trip back to the client as NaN
+
+  @odbc_e2e
+  Scenario: should accept SQL_C_BINARY infinity for SQL_REAL
```
Comment (sfc-gh-pbulawa) [inline tests/definitions/odbc/types/c_binary_conversion_to_sql_real.feature]: test -infinity

File: odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_decimal.cpp [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                 "[c_binary][conversion][sql_decimal]") {
+  SKIP_OLD_DRIVER("BD#46", "Old driver does not support SQL_C_BINARY as source for SQL_DECIMAL");
+  // Given Snowflake client is logged in
+  conn.execute("CREATE TEMPORARY TABLE t (col NUMBER(10,2))");
+
+  // When A 19-byte binary buffer containing a SQL_NUMERIC_STRUCT is bound as SQL_DECIMAL
+  auto stmt = conn.createStatement();
+  SQLRETURN ret = SQLPrepare(stmt.getHandle(), sqlchar("INSERT INTO t VALUES (?)"), SQL_NTS);
+  REQUIRE_ODBC(ret, stmt);
+  SQL_NUMERIC_STRUCT ns = make_numeric(10, 2, 1, 12345);
```
Comment (sfc-gh-pbulawa) [inline odbc_tests/tests/e2e/types/c_binary_conversion_to_sql_decimal.cpp:41]: Add a test for a numeric with negative sign

**Review by sfc-gh-pbulawa — CHANGES_REQUESTED** [review]

**Review by sfc-gh-pbulawa — APPROVED** [review]
